accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1236,18 +1248,92 @@ the `LEGACY_ERASURE` of the types.
 
 ### Const evaluation and canonicalization
 
-With unsound null checking, all generic const constructors and generic
-const literals are treated as if all type arguments passed to them were legacy
-types (both at the top level, and recursively over the structure of the
-types), regardless of whether the constructed class is defined in a legacy
-library or not, and regardless of whether the constructor invocation or
-literal occurs in a legacy library or not.  This ensures that const objects
-continue to canonicalize consistently across legacy and opted-in libraries.
+Const evaluation is modified so that legacy and opted-in instances canonicalize
+more consistently as defined below, using the operation **LEGACY_TYPE**(`T`) on
+types `T` defined as follows.
+
+- **LEGACY_TYPE**(`T`) = `T` if `T` is `dynamic`, `void`, `Null`
+- **LEGACY_TYPE**(`T`) = `T*` if `T` is `Never` or `Object`
+- **LEGACY_TYPE**(`FutureOr<T>`) = `FutureOr<S>*`
+- **LEGACY_TYPE**(`T?`) =
+  - let `S` be **LEGACY_TYPE**(`T`)
+  - if `S` is `R*` then `R?`
+  - else `S?`
+- **LEGACY_TYPE**(`T*`) = **LEGACY_TYPE**(`T`)
+- **LEGACY_TYPE**(`X extends T`) = `X*`
+- **LEGACY_TYPE**(`X & T`) =
+  - This case should not occur, since intersection types are not permitted as
+    generic arguments.
+- **LEGACY_TYPE**(`C<T0, ..., Tn>`) = `C<R0, ..., Rn>*`
+  - where `Ri` is **LEGACY_TYPE**(`Ti`)
+- **LEGACY_TYPE**(`R Function<X extends B>(S)`) = `F*`
+  - where `F = R1 Function<X extends B1>(S1)`
+  - and `R1` = **LEGACY_TYPE**(`R`)
+  - and `B1` = **LEGACY_TYPE**(`B`)
+  - and `S1` = **LEGACY_TYPE**(`S`)
+
+Note that if `T` is a normal form type, then **LEGACY_TYPE**(`T`) is also a
+normal form type.
+
+#### Type literals
+
+Two constant type literals `T1` and `T2` canonicalize to the same object if they
+are equal using the definition of runtime type equality specified above.  The
```

Comment:
> If we do decide to keep normalization to a minimum then we may be able to simply say that two types `T1` and `T2` are transformed into `**LEGACY_TYPE**(Tj)` for `j = 1` resp. `j = 2`, canonicalized to the same object iff the resulting types are identical.

I think you can avoid normalization this way, but since this is not specified as being specific to sound or unsound mode, we cannot (or at least should not) choose the legacy version of the type literal as the representative of the class.

I'm not sure what you mean by removing the arbitrary choice - the choice of instance to represent the equivalence class still seems arbitrary, regardless of whether you normalize.